rust/src/enhancers/grammar.rs

```diff
@@ -8,172 +8,266 @@
+
+            result.push(Action::Flag(action));
+            input = after_action.trim_start();
+        } else {
+            let (action, after_action) = var_action(input)
+                .with_context(|| format!("at `{input}`: failed to parse var action"))?;
+
+            result.push(Action::Var(action));
+            input = after_action.trim_start();
+        }
     }
 
-    if let Some((_, _, _, _, _, m, _, _)) = callee_matcher {
-        matchers.push(m);
+    if result.is_empty() {
+        anyhow::bail!("expected at least one action");
     }
 
-    Ok((input, matchers))
+    Ok(result)
 }
 
-fn actions(input: &str) -> IResult<&str, Vec<Action>> {
-    let max_frames = preceded(
-        tuple((tag("max-frames"), space0, char('='), space0)),
-        rule_number,
-    )
-    .map(VarAction::MaxFrames);
-
-    let min_frames = preceded(
-        tuple((tag("min-frames"), space0, char('='), space0)),
-        rule_number,
-    )
-    .map(VarAction::MinFrames);
-
-    let invert_stacktrace = preceded(
-        tuple((tag("invert-stacktrace"), space0, char('='), space0)),
-        rule_bool,
-    )
-    .map(VarAction::InvertStacktrace);
-
-    let category = preceded(tuple((tag("category"), space0, char('='), space0)), ident)
-        .map(|c| VarAction::Category(SmolStr::new(c)));
-
-    let var_action = alt((max_frames, min_frames, invert_stacktrace, category));
-
-    let flag_name = alt((
-        value(FlagActionType::Group, tag("group")),
-        value(FlagActionType::App, tag("app")),
-        value(FlagActionType::Prefix, tag("prefix")),
-        value(FlagActionType::Sentinel, tag("sentinel")),
-    ));
-    let range = opt(alt((
-        value(Range::Up, char('^')),
-        value(Range::Down, char('v')),
-    )));
-    let flag = alt((value(true, char('+')), value(false, char('-'))));
-    let flag_action =
-        tuple((range, flag, flag_name)).map(|(range, flag, ty)| FlagAction { range, flag, ty });
-
-    let action = preceded(
-        space0,
-        alt((map(flag_action, Action::Flag), map(var_action, Action::Var))),
-    );
-
-    let (input, actions) = many1(action)(input)?;
-
-    Ok((input, actions))
+fn matcher<'a>(
+    input: &'a str,
+    frame_offset: FrameOffset,
+    regex_cache: &mut RegexCache,
+) -> anyhow::Result<(Matcher, &'a str)> {
+    let input = input.trim_start();
+
+    let (negated, before_name) = if let Some(rest) = input.strip_prefix('!') {
+        (true, rest)
+    } else {
+        (false, input)
+    };
+
+    let (name, after_name) = ident(before_name)
+        .with_context(|| format!("at `{before_name}`: failed to parse matcher name"))?;
+
+    let before_arg = expect(after_name, ":")?;
+
+    let (arg, rest) = argument(before_arg)
+        .with_context(|| format!("at `{before_arg}`: failed to parse matcher argument"))?;
+
+    // TODO: support even more escapes
+    let unescaped = arg.replace("\\\\", "\\");
```

Comment:
this should be part of the `argument` parser, which then returns a `Cow<str>` depending on presence of escapes